packages/js/src/generators/typescript-sync/typescript-sync.ts

```diff
@@ -0,0 +1,315 @@
+
+      // Update project references for the runtime tsconfig
+      hasChanges =
+        updateTsConfigReferences(
+          tree,
+          runtimeTsConfigPath,
+          dependencies,
+          sourceProjectNode.data.root,
+          runtimeTsConfigFileName,
+          runtimeTsConfigFileNames
+        ) || hasChanges;
+    }
+
+    // Update project references for the tsconfig.json file
+    hasChanges =
+      updateTsConfigReferences(
+        tree,
+        sourceProjectTsconfigPath,
+        dependencies,
+        sourceProjectNode.data.root
+      ) || hasChanges;
+  }
+
+  if (hasChanges) {
+    await formatFiles(tree);
+
+    return 'Based on the workspace project graph, some TypeScript configuration files are missing project references to the projects they depend on.';
+  }
+}
+
+export default syncGenerator;
+
+function updateTsConfigReferences(
+  tree: Tree,
+  tsConfigPath: string,
+  dependencies: ProjectGraphProjectNode[],
+  projectRoot: string,
+  runtimeTsConfigFileName?: string,
+  possibleRuntimeTsConfigFileNames?: string[]
+): boolean {
+  const tsConfig = readJson<Tsconfig>(tree, tsConfigPath);
+  // We have at least one dependency so we can safely set it to an empty array if not already set
+  const references = [];
+  const originalReferencesSet = new Set();
+  const newReferencesSet = new Set();
+  for (const ref of tsConfig.references ?? []) {
+    const normalizedPath = normalizeReferencePath(ref.path);
+    originalReferencesSet.add(normalizedPath);
+    if (!ref.path.startsWith('../')) {
+      // we only keep the internal references, the external ones are dictated by the project dependencies
+      references.push(ref);
+      newReferencesSet.add(normalizeReferencePath(ref.path));
+    }
+  }
+
+  let hasChanges = false;
+  for (const dep of dependencies) {
+    // Ensure the project reference for the target is set
+    let referencePath = dep.data.root;
+    if (runtimeTsConfigFileName) {
+      const runtimeTsConfigPath = joinPathFragments(
+        dep.data.root,
+        runtimeTsConfigFileName
+      );
+      if (tree.exists(runtimeTsConfigPath)) {
+        referencePath = runtimeTsConfigPath;
+      } else {
+        // Check for other possible runtime tsconfig file names
+        // TODO(leo): should we check if there are more than one runtime tsconfig files and throw an error?
+        for (const possibleRuntimeTsConfigFileName of possibleRuntimeTsConfigFileNames ??
+          []) {
+          const possibleRuntimeTsConfigPath = joinPathFragments(
+            dep.data.root,
+            possibleRuntimeTsConfigFileName
+          );
+          if (tree.exists(possibleRuntimeTsConfigPath)) {
+            referencePath = possibleRuntimeTsConfigPath;
+            break;
+          }
+        }
+      }
+    }
+    const relativePathToTargetRoot = relative(projectRoot, referencePath);
+    if (!newReferencesSet.has(relativePathToTargetRoot)) {
+      newReferencesSet.add(relativePathToTargetRoot);
```

Comment:
Don't delete if the file exists.

Reply:
It's a bit more complex than that and I'll address it in a separate PR after this one lands.